feat(dart/catalyst_cardano_serialization): initial implementation of dynamic coin selection algorithm

ilap opened the pull request:
# Description

This PR introduces the initial implementation of a coin selection algorithm in the `TransactionBuilder`. The algorithm automatically selects the set of unspent transaction outputs (UTXOs) for constructing a transaction. 
It aims to minimise transaction fees and improve overall efficiency.

Key changes include:
- Implemented the dynamic coin selection logic within the TransactionBuilder.

## Related Issue(s)

Resolves #1520

## Breaking Changes

There are no breaking changes in this pull request.

## Screenshots

N/A

## Related Pull Requests

N/A

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules

dtscalac commented:
@ilap thank you for the contribution, will review it later today.

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:
```diff
@@ -0,0 +1,61 @@
+  const config = SelectionUtils.defaultConfig;
+  final changeAddress = SelectionUtils.randomAddress();
+
+  group('TransactionBuilder Coin Selection Tests', () {
```
Often when referring to classes we can inline the type so that when the class is renamed the test is also updated.
```suggestion
  group('$TransactionBuilder Coin Selection Tests', () {
```

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:
```diff
@@ -0,0 +1,61 @@
+              anyOf([
+                returnsNormally,
+                throwsA(isA<MaxTxSizeExceededException>()),
+                //isNot(isA<InsufficientUtxoBalanceException>()),
```
Please remove commented out code

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:
```diff
@@ -0,0 +1,61 @@
+              maxOutputs: 10,
+            );
+
+            //print('#${i++}: Utxos: $utxoCount, outputs: ${outputs.length}.');
```
Remove commented out code

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:
```diff
@@ -0,0 +1,61 @@
+    });
+
+    group('Property-based tests for coin selection', () {
+      //var i = 0;
```
Remove commented out code

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:
```diff
@@ -0,0 +1,61 @@
+
+    group('Property-based tests for coin selection', () {
+      //var i = 0;
+      property(
+          'Coin selection should maintain accounting properties for '
+          'various UTxO and output combinations', () {
```
While I value fuzzy tests it can be dangerous too. Consider to add tests that use hardcoded data and cover minimal, maximal and "in the middle" cases.

dtscalac replied:
Fuzzy tests can stay, just want to have a stable set of tests that validate some known cases.

ilap replied:
This initial property-based test should be stable, as it's for just checking the insufficient utxo balances, which should not happen if the utxo accounting properties are met.

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,403 @@
+/// operations for testing purposes.
+///
+/// This class is used only in testing and resides under the `./test` folder.
+sealed class SelectionUtils {
```
Does the class need to be extended? If not consider making it final

```suggestion
final class SelectionUtils {
```

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,403 @@
+  /// - A random [ShelleyAddress].
+  static ShelleyAddress randomAddress({
+    bool isBase = true,
+    bool isMainnet = false,
```
This should use `NetworkId` enum

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,403 @@
+  static List<ShelleyAddress> randomAddresses({
+    required int count,
+    bool isBase = true,
+    bool isMainnet = false,
```
`NetworkId` enum

dtscalac commented:
The spell checker is complaining about these issues:

```
     +check-spelling *failed* | Repeating the failure error...
     +check-spelling *failed* | --> RUN cspell-cli --quiet lint . --dot
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart:16:24 - Unknown word (prioritise)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart:56:45 - Unknown word (prioritise)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart:778:38 - Unknown word (lentgh)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart:844:44 - Unknown word (multiassets)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/types.dart:31:52 - Unknown word (prioritised)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/pubspec.yaml:29:3 - Unknown word (kiri)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:2:17 - Unknown word (kiri)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/test/builders/input_builder_test.dart:2:28 - Unknown word (kiri)
     +check-spelling *failed* | catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:376:28 - Unknown word (normalise)
     +check-spelling *failed* | ERROR /var/folders/g1/55kjtgl504v466583q_cg4440000gp/T/earthly-git2873568597/Earthfile:20:4
     +check-spelling *failed* |       The command
     +check-spelling *failed* |           RUN cspell-cli --quiet lint . --dot
     +check-spelling *failed* |       did not complete successfully. Exit code 1

```

The ones that are not typos should be added to `.config/dictionaries/project.dic` so that the spell-checker ignores them.

ilap commented:
Thx, will go through all of them soon.

dtscalac commented:
@ilap 

The changes from this branch are causing the following error in one of the dependent packages (but I think it's an issue with the serializing package):

![Screenshot 2025-01-29 at 10 06 09](https://github.com/user-attachments/assets/262dcca8-e8c9-40c3-aa36-6eb672d1b1ce)

To reproduce:
1. cd catalyst_voices/packages/libs/catalyst_cardano/catalyst_cardano/example
2. flutter run \                                                             
--web-header Cross-Origin-Opener-Policy=same-origin \
--web-header Cross-Origin-Embedder-Policy=require-corp \
--target lib/main.dart \
-d chrome
3. connect any wallet (test-net preferred)
4. attempt to `sign & submit tx`

ilap commented:
> On current `main` branch the issue is not reproducible.

Hey @dtscalac, thanks for the heads-up!

My latest commit should fix this. The main issue was that the output fee calculation (`feeForOutput`) method always returned 0, so I fixed that. I also cleaned up balance handling and tweaked asset mapping.

Can you give it another shot and see if everything works now?

ilap commented:
Hey @dtscalac , Here are some suggestions:
- Deprecate `buildFake` and `vkeysCount` in `WitnessBuilder`.
- Deprecate the following methods in `TransactionBuilder`:
  - `withChangeAddressIfNeeded`
  - `_withChangeAddressIfNeededWithMultiAssets`
  - `_withChangeAddressIfNeededWithoutMultiAssets`
  - `_willAddingAssetMakeOutputOverflow`
  - `_packNftsForChange`

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/types.dart:
```diff
@@ -0,0 +1,185 @@
+/// transaction's outputs, while also calculating necessary change.
+abstract interface class CoinSelector {
+  /// The predefined [PolicyId] for ADA.
+  static PolicyId adaPolicy = PolicyId('');
```
Does this need to be a variable? Do we intend to reassign it at some point? Can we consider to make it final/const?

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,437 @@
+    bool withTokens = false,
+  }) {
+    final pids = List.generate(10, (int index) {
+      return PolicyId(randomHexString(28));
```
Consider to extract the `28` to a const, preferably as static field in `PolicyId`

ilap replied:
done

dtscalac replied:
Can we reference the const we extracted in here?

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,437 @@
+      count,
+      (i) => TransactionUnspentOutput(
+        input: TransactionInput(
+          transactionId: TransactionHash.fromHex(randomHexString(32)),
```
```suggestion
          transactionId: TransactionHash.fromHex(randomHexString(TransactionHash.hashLength)),
```

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,437 @@
+      ShelleyAddress(
+        [
+          _getMockAddressHrp(isBase: isBase, networkId: networkId),
+          ...randomBytes(isBase ? 56 : 28),
```
Consider to extract `56` and `28` as named constants

dtscalac replied:
Preferably in a class related to what they mean

ilap replied:
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart:
```diff
@@ -525,14 +657,16 @@ final class TransactionBuilder extends Equatable {
 
-  TransactionBuilder _copyWith({
-    List<ShelleyMultiAssetTransactionOutput>? outputs,
+  /// CopyWidth
```
Consider to improve this comment or lets remove it, doesn't add anything valuable at the moment

ilap replied:
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/strategies/random_selection_strategy.dart:
```diff
@@ -0,0 +1,16 @@
+class RandomSelectionStrategy implements CoinSelectionStrategy {
+  final Random _random = Random();
+
+  @override
+  @override
+  void apply(AssetsGroup assetsGroup) {
```
```suggestion
  @override
  void apply(AssetsGroup assetsGroup) {
```

ilap replied:
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart:
```diff
@@ -0,0 +1,296 @@
+  final CoinSelectionStrategy selectionStrategy;
+
+  /// Creates an [InputBuilder] with the given [selectionStrategy].
+  InputBuilder({required this.selectionStrategy});
```
Consider to add a `const `constructor to `CoinSelector` so that we can make `InputBuilder` constructor `const` too.

ilap replied:
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart:
```diff
@@ -0,0 +1,296 @@
+  ) {
+    if (remainingBalance.isZero) return ([], const Coin(0));
+    if (builder.changeAddress == null) {
+      throw Exception('Change address required for non-zero balance.');
```
I'd change it to an `Error` because it's programmer's mistake to not pass the change address.

```suggestion
      throw ArgumentError('Change address required for non-zero balance.');
```

ilap replied:
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/input_builder.dart:
```diff
@@ -0,0 +1,296 @@
+
+      // Determine if change should be calculated for the current asset group.
+      final shouldCalculateChange = assetId == CoinSelector.adaAssetId ||
+          assetGroups[groupCount - 2].key == assetId;
```
Do we always expect the ADA assets to be at the `groupCount - 2` index? I wasn't able to identify which lines guarantee that in `buildAssetGroups`, is it these ones:

```
  return assetMap.entries.toList()
      ..sort((a, b) => b.key.$1.hash.compareTo(a.key.$1.hash));
```
?

ilap replied:
Sorry, this is a bit misleading. The ADA assets (which contains all the inputs, as all input contains coin) must be the last in the group and the hash is '' for ADA, meaning reverse-lexicographically  comparing the  the pids  (b.compareTo(a)) satisfy this requirement.
The groupCount - 2, means that there could be multiple or zero non-ada assetId in the group e.g.  [adaAsset] or [assetId1, assetId2, adaAssetId]  etc. and - 2 means the token before ADA (as dart uses short-circuit, that means if it is not adaAssetId then || assetGroups[groupCount -2] is always points to an existing assetId before ADA). It's for the case when all tokens have been processed and we can calculate the change now. As it's should not be calculated until all target/payment tokens are processed (no point of doing it). hope it makes sense

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/types.dart:
```diff
@@ -0,0 +1,185 @@
+
+/// Represents the valid range for coin values as a tuple of minimum and
+/// maximum [Coin] values.
+typedef ValidRange = (Coin, Coin);
```
For all the typedefs defined in here I would add a name alongside the type to all function parameters.

I.e.

```suggestion
typedef ValidRange = (Coin minInclusive, Coin maxInclusive);
```

dtscalac replied:
as not every parameter goal can be deducted just from reading the code here

ilap replied:
done

dtscalac commented:
> Hey @dtscalac , Here are some suggestions:
> 
>     * Deprecate `buildFake` and `vkeysCount` in `WitnessBuilder`.
> 
>     * Deprecate the following methods in `TransactionBuilder`:
>       
>       * `withChangeAddressIfNeeded`
>       * `_withChangeAddressIfNeededWithMultiAssets`
>       * `_withChangeAddressIfNeededWithoutMultiAssets`
>       * `_willAddingAssetMakeOutputOverflow`
>       * `_packNftsForChange`

Yes, makes sense!. At this stage of the package lifecycle I'd recommend making them as breaking changes and not maintaining deprecated code. The package must retain it's functionality and for every feature we remove there should be a replacement of some sort.

dtscalac commented:
> My latest commit should fix this. The main issue was that the output fee calculation (`feeForOutput`) method always returned 0, so I fixed that. I also cleaned up balance handling and tweaked asset mapping.
> 
> Can you give it another shot and see if everything works now?

Yes, now the issue has been fixed, thanks ❤️

ilap commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/types.dart:
```diff
@@ -0,0 +1,185 @@
+  /// that must be selected for inclusion in a transaction. It defaults to 0.
+  ///
+  ///
+  static const minInputs = 0;
```
done

ilap commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/exceptions.dart:
```diff
@@ -25,14 +25,32 @@ final class MaxTxSizeExceededException extends Equatable implements Exception {
+/// Exception thrown when the maximum number of inputs is exceeded.
+final class MaximumInputExceededException extends Equatable
+    implements Exception {
+  /// The maximum nr. of allowed inputs.
```
done

ilap commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/test/test_utils/selection_utils.dart:
```diff
@@ -0,0 +1,437 @@
+  /// Returns:
+  /// - A random hexadecimal string.
+  static String randomHexString(int length) =>
+      hex.encoder.convert(randomBytes(length));
```
done

dtscalac commented on catalyst_voices/packages/libs/catalyst_cardano_serialization/lib/src/builders/types.dart:
```diff
@@ -0,0 +1,185 @@
+/// transaction's outputs, while also calculating necessary change.
+abstract interface class CoinSelector {
+  /// The predefined [PolicyId] for ADA.
+  static PolicyId adaPolicy = PolicyId('');
```
@ilap Did you make any changes here? Or if you think this is correct please let me know why :)

ilap replied:
Sorry, I missed that.  The others are final there. Fixing now.

dtscalac reviewed (approved):
LGTM

@minikin Can you help with merging the PR? I've verified it successfully!

dtscalac commented:
@ilap 
Sure, new versions will be published in https://github.com/input-output-hk/catalyst-voices/pull/1781
